lean4-infoview/src/infoview/info.tsx

```diff
@@ -0,0 +1,266 @@
+                    <div className="ml1">
+                        <MessagesAtFile uri={pos.uri} messages={messages}/>
+                    </div>
+                </Details>}
+            {nothingToShow && (
+                isPaused ?
+                    <span>Updating is paused. <a className="link pointer dim" onClick={e => { e.preventDefault(); triggerDisplayUpdate(); }}>Refresh</a> or <a className="link pointer dim" onClick={e => { e.preventDefault(); setPaused(false); }}>resume updating</a> to see information.</span> :
+                    status === 'ready' ? 'No info found.' : 'Loading...')}
+        </div>
+    </Details>
+    );
+}
+
+function isLoading(ts: LeanFileProgressParams, p: DocumentPosition): boolean {
+    return ts.processing.some(i => RangeHelpers.contains(i.range, p));
+}
+
+/**
+ * returns function that triggers `cb`
+ * - but only `ms` milliseconds after the first call
+ * - and not more often than once every `ms` milliseconds
+ */
+function useDelayedThrottled(ms: number, cb: () => Promise<void>): () => Promise<void> {
+    const waiting = React.useRef<boolean>(false);
+    const callbackRef = React.useRef<() => Promise<void>>();
+    callbackRef.current = cb;
+    return async () => {
+        if (!waiting.current) {
+            waiting.current = true;
+            let promise = new Promise((resolved, rejected) => {
+                setTimeout(() => {
+                    waiting.current = false;
+                    callbackRef.current!().then(resolved, rejected);
+                }, ms);
+            });
+            await promise;
+        }
+    };
+}
+
+/**
+ * Note: in the cursor view, we have to keep the cursor position as part of the component state
+ * to avoid flickering when the cursor moved. Otherwise, the component is re-initialised and the
+ * goal states reset to `undefined` on cursor moves.
+ */
+export type InfoProps = InfoPinnable & { pos?: DocumentPosition };
+
+/** Fetches info from the server and renders an {@link InfoDisplay}. */
+export function Info(props: InfoProps) {
+    const ec = React.useContext(EditorContext);
+
+    // Note: `kind` may not change throughout the lifetime of an `Info` component,
+    // otherwise the hooks will differ.
+    const pos = props.kind === 'pin' ?
+        props.pos! :
+        (() => {
+            const [curLoc, setCurLoc] = React.useState<Location>(ec.events.changedCursorLocation.current!);
+            useEvent(ec.events.changedCursorLocation, setCurLoc, []);
+            return { uri: curLoc.uri, ...curLoc.range.start };
+        })();
+
+    const [status, setStatus] = React.useState<InfoStatus>('loading');
+    const [goal, setGoal] = React.useState<PlainGoal>();
+    const [termGoal, setTermGoal] = React.useState<PlainTermGoal>();
+    const [error, setError] = React.useState<string>();
+
+    const messages = useMessagesFor(pos);
+    const [serverIsProcessing, _] = useServerNotificationState(
+        '$/lean/fileProgress',
+        false,
```

Comment:
I think this is the same problem as with diagnostics, namely that the editor needs to re-send the notification it received from the server on infoview startup. I'm not seeing a nice way to solve this within LSP, so I'll do what the current version does which is re-send diagnostics and progress on infoview startup.

Reply:
It will also be initialized with the wrong value of the Info is opened while the extension is running (e.g. when pinning).  The false here should really be something like `isLoading(currentFileProgressParams, pos)`.

Reply:
I think you're right, this should be top-level infoview state like the diagnostics rather than per-component state.